### PR TITLE
Strong typing for Stellar

### LINKED ISSUE #2: Test each and every operation in Stellar
Some operations are not tested. I belive those are:

- StellarAllowTrustOp
- StellarPathPaymentOp
- StellarChangeTrustOp
- StellarCreatePassiveOfferOp
- StellarManageDataOp
- StellarManageOfferOp
- StellarPathPaymentOp

Could avoid https://github.com/trezor/connect/issues/382

### LINKED ISSUE #1699: Stellar: invalid signature if timebounds are set to 0/0
Trezor (both T1 and TT) conflate timebounds set to 0/0 with timebounds unset, and produce a signature for "timebounds unset". This is of course invalid if the incoming parameters are "set to 0/0"

---

### Original report:

**Describe the bug**
On [Stellar Laboratory](https://laboratory.stellar.org/) created a payment transaction without the upper time bound. The transaction was shown correctly on the device and signed. When validating the signature on the Laboratory, it was invalid. The signature validates correctly if the upper time bound is set.

**Firmware version and revision**
Trezor One v1.10.1

**Desktop/smartphone setup (please complete the following information):**
 - Wallet software: Trezor Wallet
 - OS: Windows and MacOS (Trezor Bridge v2.0.27)
 - Browser: Chrome v91.0.4472.124

**To Reproduce**
1. Go to [Laboratory](https://laboratory.stellar.org/)
2. Select `public` network (top right corner)
2. Click on "Build Transaction" tab and fill in:
- Source Account
- Transaction Sequence Number (click on Fetch... button)
- Select Payment in Operation Type dropdown
- Destination
- Asset `native`
- Amount
3. Click "Sign in Transaction Signer" at the bottom
4. Scroll to the bottom of the page and click "Sign with Trezor"
5. Follow the Trezor instructions
6. Should see "Success!" message that the signature was added
7. Click on "Submit in Transaction Submitter" at the bottom of the page
8. Scroll to the bottom of the page to view checked signatures

image

**Expected behavior**
The upper time bound should not be required to validate the signature.

### PR DESCRIPTION
**Pull request by matejcik**
Changes most fields in Stellar protobuf to `required`. This mostly matches the Stellar XDR, so it should not introduce incompatibilities on caller sides.

The notable exception to "mostly matches" are the timebounds fields. In Stellar XDR, they are optional, and in the past, we would treat them as such. This PR makes timebounds required, and so rejects transactions without timebounds.
Per https://github.com/stellar/js-stellar-base/issues/438 omitting timebounds is strongly discouraged by documentation and existing implementations, so I feel that we do not need to support this feature.
Incidentally, this fixes #1699 

Several fields (`memo_type`, `signer_type`, `asset.type`) now use enums. There is another incompatibility: `StellarAssetType` is renamed to `StellarAsset`, so that `StellarAssetType` can be used as the name of the enum.

Enum values are binary-compatible with the previously existing fields.

ping @szymonlesisz or @matejkriz for the above, which require modifications in Connect:
1. if the incoming data structure does not have `timebounds`, it must be rejected
2. rename `StellarAssetType` to `StellarAsset` as appropriate. Or let me know if this is a bigger problem and we'll figure out some other way to name this.

---

all testcases have been regenerated with timebounds, and signatures checked against Stellar Laboratory. I renamed the testcases, which unfortunately breaks the UI diff.
fixes #2

### CODE REVIEW

File: common/protob/messages-stellar.proto
Comment (overcat) [inline common/protob/messages-stellar.proto:205]: I don’t think it should be a bool value here.

https://github.com/stellar/stellar-core/blob/master/src/xdr/Stellar-transaction.x#L269

Comment (matejcik) [reply common/protob/messages-stellar.proto:205]: good catch, but it's treated in code as bool anyway:
https://github.com/trezor/trezor-firmware/blob/master/core/src/apps/stellar/operations/serialize.py#L30
https://github.com/trezor/trezor-firmware/blob/master/legacy/firmware/stellar.c#L989

i'm wondering why it isn't defined as an enum in the XDR?

Comment (overcat) [reply common/protob/messages-stellar.proto:205]: It was indeed bool before, but it has changed with the update of the protocol. https://github.com/stellar/stellar-core/commit/29669910a7ef57db6b21dc86ee207fda3cbc1c39#diff-11f6a6577ee4d25e4eb1b1834bd71486cc73c26ad4eda68345286a27f5d74307R236

> i'm wondering why it isn't defined as an enum in the XDR?

I don't know why, but my personal guess is to support bitwise operators?

Comment (overcat) [reply common/protob/messages-stellar.proto:205]: > good catch, but it's treated in code as bool anyway:
> https://github.com/trezor/trezor-firmware/blob/master/core/src/apps/stellar/operations/serialize.py#L30
> https://github.com/trezor/trezor-firmware/blob/master/legacy/firmware/stellar.c#L989

Do you plan to modify it in this PR? At present, if passing `is_authorized` is not 0 or 1, the user cannot get the correct signature.

Comment (matejcik) [reply common/protob/messages-stellar.proto:205]: > I don't know why, but my personal guess is to support bitwise operators?

originally yes, apparently, but the comment is pretty explicit and there's even a [review comment](https://github.com/stellar/stellar-core/pull/2881#discussion_r560476438) about it

i'll probably do an enum in our protobuf anyway, but it would be nicer to be in line with upstream about it

Comment (matejcik) [reply common/protob/messages-stellar.proto:205]: As mentioned in your PR, I won't be changing this for an operation that is deprecated anyway. We'll leave that to if/when we implement the additional Stellar operations.

**Comment by overcat** [comment]
Hi @matejcik, any updates here? If this PR can be merged, I can resolve #1691 and #1698 on this basis.

**Comment by matejcik** [comment]
This is waiting for an internal review and presumably will be merged in September.

Please hold off on implementing the mentioned issues, it is currently unclear if we would accept such PRs.

**Comment by overcat** [comment]
> This is waiting for an internal review and presumably will be merged in September.
> 
> Please hold off on implementing the mentioned issues, it is currently unclear if we would accept such PRs.

Thank you for your quick reply, I will discuss with you before implementation.

**Comment by JFWooten4** [comment]
How to update our firmware should this merge to main?

File: core/src/apps/stellar/layout.py
Comment (mmilata) [inline core/src/apps/stellar/layout.py:60]: I know it's a change wrt to original code, but shouldn't we show human readable time here like T1 does?

**Review by mmilata — APPROVED** [review]
LGTM! UI diff ack. Thanks for getting rid of `confirm_timebounds_stellar`.

File: core/src/apps/stellar/layout.py
Comment (matejcik) [reply core/src/apps/stellar/layout.py:60]: we can't currently because we don't have epoch-to-datetime decoding.

and we will get it with uPy 1.14 (see #1535) so this will be fixed then

File: common/protob/messages-stellar.proto
Comment (matejcik) [reply common/protob/messages-stellar.proto:9]: correct

File: core/src/trezor/crypto/base32.py
Comment (matejcik) [reply core/src/trezor/crypto/base32.py]: fixed in rebase

**Comment by matejcik** [comment]
rebased, fixed conflict with master.

this is now waiting for input from Connect side

**Comment by matejcik** [comment]
TODO:
* [ ] fix tests
* [ ] add changelogs

**Comment by bosomt** [comment]
@matejcik anything that can/need to be checked on our side ?

**Comment by matejcik** [comment]
is Stellar Account Viewer in your scope? if yes, then general operation of Stellar

**Comment by matejcik** [comment]
in particular please check that https://github.com/trezor/trezor-firmware/issues/1699 is fixed

**Comment by bosomt** [comment]
@matejcik Stellar is not in our scope, 
but of course we will keep it in needs QA column and we will test it during release testing

**Pull request merged**
